File: specs/TruexAdEvent.ts
/**
 * Describes events emitted from the truex ad renderer as the interactive ad progresses through its life cycle.
 */
export enum TruexAdEvent {
  AdStarted = 'adStarted',
  AdDisplayed = 'adDisplayed',
  AdCompleted = 'adCompleted',
  AdError = 'adError',
  NoAdsAvailable = 'noAdsAvailable',
  AdFreePod = 'adFreePod',
  AdFetchCompleted = 'adFetchCompleted',
  UserCancelStream = 'userCancelStream',
  OptIn = 'optIn',
  OptOut = 'optOut',
  SkipCardShown = 'skipCardShown',
  UserCancel = 'userCancel',
  XtendedViewStarted = 'xtendedViewStarted',
  PopupWebsite = "popupWebsite"
}

export type AdEventData = undefined | null | Record<string, null | boolean | number | string>;
export type AdEventHandler = (event: TruexAdEvent, data?: AdEventData) => void;

export function isCompletionEvent(event: TruexAdEvent) {
  switch (event) {
    case TruexAdEvent.NoAdsAvailable:
    case TruexAdEvent.AdCompleted:
    case TruexAdEvent.AdError:
    case TruexAdEvent.UserCancelStream:
      return true;
  }
  return false;
}

File: android/app/src/main/java/com/tinkerbell/tar/TruexAdEvent.java
package com.tinkerbell.tar;

public enum TruexAdEvent {
  AD_FETCH_COMPLETED("adFetchCompleted"),
  AD_STARTED("adStarted"),
  AD_DISPLAYED("adDisplayed"),
  AD_COMPLETED("adCompleted"),
  AD_ERROR("adError"),
  NO_ADS_AVAILABLE("noAdsAvailable"),
  AD_FREE_POD("adFreePod"),
  USER_CANCEL_STREAM("userCancelStream"),
  OPT_IN("optIn"),
  OPT_OUT("optOut"),
  SKIP_CARD_SHOWN("skipCardShown"),
  USER_CANCEL("userCancel"),
  VIDEO_EVENT("videoEvent"),
  POPUP_WEBSITE("popupWebsite"),
  XTENDED_VIEW_STARTED("xtendedViewStarted");

  private final String name;

  TruexAdEvent(String name) {
    this.name = name;
  }

  public String toString() {
    return this.name;
  }

  static public TruexAdEvent toEvent(String eventName) {
    for(TruexAdEvent event : TruexAdEvent.values()) {
      if (event.name.equals(eventName)) return event;
    }
    throw new IllegalArgumentException("ad event not found: " + eventName);
  }
}
